<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">


<link rel="stylesheet" href="/lib/font-awesome/css/all.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"biomarble.github.io","root":"/","scheme":"Pisces","version":"7.8.0","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":false,"show_result":false,"style":null},"back2top":{"enable":true,"sidebar":false,"scrollpercent":false},"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":false,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}}};
  </script>

  <meta name="description" content="Quaerendo invenietis觅之，自有所获">
<meta property="og:type" content="website">
<meta property="og:title" content="BioLog">
<meta property="og:url" content="http://biomarble.github.io/index.html">
<meta property="og:site_name" content="BioLog">
<meta property="og:description" content="Quaerendo invenietis觅之，自有所获">
<meta property="og:locale" content="zh_CN">
<meta property="article:author" content="Marble">
<meta property="article:tag" content="Bioinformatics">
<meta name="twitter:card" content="summary">

<link rel="canonical" href="http://biomarble.github.io/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : true,
    isPost : false,
    lang   : 'zh-CN'
  };
</script>

  <title>BioLog</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">BioLog</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">Yet Another Bioinformatics Blog</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="main-menu menu">
        <li class="menu-item menu-item-home">

    <a href="/" rel="section"><i class="fa fa-home fa-fw"></i>首页</a>

  </li>
        <li class="menu-item menu-item-about">

    <a href="/about/" rel="section"><i class="fa fa-user fa-fw"></i>关于</a>

  </li>
        <li class="menu-item menu-item-tags">

    <a href="/tags/" rel="section"><i class="fa fa-tags fa-fw"></i>标签<span class="badge">6</span></a>

  </li>
        <li class="menu-item menu-item-categories">

    <a href="/categories/" rel="section"><i class="fa fa-th fa-fw"></i>分类<span class="badge">3</span></a>

  </li>
        <li class="menu-item menu-item-archives">

    <a href="/archives/" rel="section"><i class="fa fa-archive fa-fw"></i>归档<span class="badge">3</span></a>

  </li>
        <li class="menu-item menu-item-software">

    <a href="/software/" rel="section"><i class="fa fa-box fa-fw"></i>软件</a>

  </li>
  </ul>
</nav>




</div>
    </header>

    
  <div class="back-to-top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content index posts-expand">
            
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://biomarble.github.io/2020/03/20/scRNA-seq%E4%B8%AD%E7%9A%84Doublet-Multiplet%E7%9A%84%E6%A3%80%E6%B5%8B%E6%96%B9%E6%B3%95/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://img.discogs.com/w9vJTN3wwe3nTq7628A9o9QFb6k=/fit-in/600x597/filters:strip_icc():format(jpeg):mode_rgb():quality(90)/discogs-images/R-1752295-1430835599-7152.jpeg.jpg">
      <meta itemprop="name" content="Marble">
      <meta itemprop="description" content="Quaerendo invenietis<br>觅之，自有所获">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="BioLog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2020/03/20/scRNA-seq%E4%B8%AD%E7%9A%84Doublet-Multiplet%E7%9A%84%E6%A3%80%E6%B5%8B%E6%96%B9%E6%B3%95/" class="post-title-link" itemprop="url">scRNA-seq中的Doublet/Multiplet的检测方法</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2020-03-20 18:35:51" itemprop="dateCreated datePublished" datetime="2020-03-20T18:35:51+08:00">2020-03-20</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">更新于</span>
                <time title="修改时间：2021-04-15 22:43:19" itemprop="dateModified" datetime="2021-04-15T22:43:19+08:00">2021-04-15</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E5%8D%95%E7%BB%86%E8%83%9E%E6%B5%8B%E5%BA%8F/" itemprop="url" rel="index"><span itemprop="name">单细胞测序</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h1 id="Doublet-Multiplet-的检测方法"><a href="#Doublet-Multiplet-的检测方法" class="headerlink" title="Doublet/Multiplet 的检测方法"></a>Doublet/Multiplet 的检测方法</h1><h2 id="Seurat"><a href="#Seurat" class="headerlink" title="Seurat"></a>Seurat</h2><p>Seurat软件包是SingleCell的基本分析软件工具包。<br>该软件包利用一种特殊的处理方式（oligo-tagged antibodies against ubuquitously expressed surface proteins to place a “sample barcode” on each single cell）得到的结果，对同一个液滴存在多个细胞的情况进行demultiplex，具体方法请单击： <a target="_blank" rel="noopener" href="https://satijalab.org/seurat/v3.1/hashing_vignette.html">vignette</a>.</p>
<p>ps: 一般的scRNA-seq方式得到的结果无法使用该方法计算</p>
<h2 id="DoubletFinder"><a href="#DoubletFinder" class="headerlink" title="DoubletFinder"></a><a target="_blank" rel="noopener" href="https://github.com/chris-mcginnis-ucsf/DoubletFinder">DoubletFinder</a></h2><p>DoubletFinder软件可以通过以下四个步骤推断Doublet：</p>
<ol>
<li>从现有的scRNA数据中产生人工的doublet，方法是随机挑选细胞对的基因表达求平均（pN）</li>
<li>预处理合并后的真假混合数据</li>
<li>对混合数据计算PCA，利用pc计算每个细胞中包含doublet成分的比例</li>
<li>对细胞进行排序</li>
</ol>
<blockquote>
<p>(1) Generate artificial doublets from existing scRNA-seq data</p>
<p>(2) Pre-process merged real-artificial data</p>
<p>(3) Perform PCA and use the PC distance matrix to find each cell’s proportion of artificial k nearest neighbors (pANN)</p>
<p>(4) Rank order and threshold pANN values according to the expected number of doublets</p>
</blockquote>
<p><img src="https://github.com/chris-mcginnis-ucsf/DoubletFinder/raw/master/DF.screenshots/DoubletFinderOverview.png" alt="DoubletFinder"></p>
<p>使用方法：</p>
<p>经测试，该程序适用seurat2版本的object，因此需要先按照seurat2</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">conda create -n  doubletfinder   r-seurat=2 r-modes=0.7.0 r-matrix=1.2.14   r-fields=9.6  r-devtools</span><br><span class="line">...</span><br><span class="line"><span class="comment">#####未完，待测试</span></span><br></pre></td></tr></table></figure>





<h2 id="DoubletDetection"><a href="#DoubletDetection" class="headerlink" title="DoubletDetection"></a><a target="_blank" rel="noopener" href="https://github.com/JonathanShor/DoubletDetection">DoubletDetection</a></h2><p>DoubletDetection 是一个使用Python3语言的scRNA-seq  Doublet检测软件。</p>
<p>使用方法：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#安装DoubletDetection，需事先准备好python3环境</span></span><br><span class="line"><span class="comment">#如果没有python3环境，可以使用conda  create  -n py3 python=3 创建</span></span><br><span class="line">git <span class="built_in">clone</span> https://github.com/JonathanShor/DoubletDetection.git</span><br><span class="line"><span class="built_in">cd</span> DoubletDetection</span><br><span class="line">pip3 install .</span><br></pre></td></tr></table></figure>

<p>DoubletDetection 运行代码：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> doubletdetection</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> doubletdetection</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line">matrix_path = <span class="string">&#x27;bmy1/matrix.mtx&#x27;</span>   <span class="comment"># 为cellranger结果文件夹中的matrix.mtx文件，需解压出来</span></span><br><span class="line">raw_counts = doubletdetection.load_mtx(matrix_path)</span><br><span class="line">zero_genes = (np.<span class="built_in">sum</span>(raw_counts, axis=<span class="number">0</span>) == <span class="number">0</span>).A.ravel() <span class="comment">#去除表达0的基因</span></span><br><span class="line">raw_counts = raw_counts[:, ~zero_genes]  </span><br><span class="line"></span><br><span class="line">clf = doubletdetection.BoostClassifier(n_iters=<span class="number">500</span>, use_phenograph=<span class="literal">False</span>, standard_scaling=<span class="literal">True</span>)   <span class="comment"># n_iters为迭代次数，当然越大越好，但是下一步计算时间会更长</span></span><br><span class="line">doublets = clf.fit(raw_counts).predict(p_thresh=<span class="number">1e-16</span>, voter_thresh=<span class="number">0.5</span>) </span><br><span class="line">np.savetxt(<span class="string">&#x27;out1.txt&#x27;</span>, doublets, fmt=<span class="string">&quot;%d&quot;</span>, delimiter=<span class="string">&#x27;,&#x27;</span>)  <span class="comment">#输出æ终预测结果到out1.txt文件中</span></span><br><span class="line"><span class="comment"># 结果文件中</span></span><br><span class="line"><span class="comment"># 1 :doublet</span></span><br><span class="line"><span class="comment"># 0 :singlet</span></span><br><span class="line"><span class="comment"># np.nan :ambiguous cell，无法判断</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#以下为绘图（非必须）</span></span><br><span class="line"><span class="comment">#注意，以下步骤可能需要X11环境（即纯terminal可能报错）</span></span><br><span class="line"><span class="comment">#convergence_test.pdf 为预测的doublet数随着迭代次数变化的收敛曲线</span></span><br><span class="line"><span class="comment">#umap_test.pdf   为注明了doublet的tsne图（？）</span></span><br><span class="line"><span class="comment">#threshold_test.pdf 为不同参数阈值组合下预测的doublet数目的热图</span></span><br><span class="line">f = doubletdetection.plot.convergence(clf, save=<span class="string">&#x27;convergence_test.pdf&#x27;</span>, show=<span class="literal">True</span>, p_thresh=<span class="number">1e-16</span>, voter_thresh=<span class="number">0.5</span>)</span><br><span class="line">f2, umap_coords = doubletdetection.plot.umap_plot(raw_counts, doublets, random_state=<span class="number">1</span>,</span><br><span class="line">                                                               save=<span class="string">&#x27;umap_test.pdf&#x27;</span>, show=<span class="literal">True</span>)</span><br><span class="line">f3 = doubletdetection.plot.threshold(clf, save=<span class="string">&#x27;threshold_test.pdf&#x27;</span>, show=<span class="literal">True</span>, p_step=<span class="number">6</span>)</span><br></pre></td></tr></table></figure>

<p>以上代码可以再python3环境中一句句运行，也 可以写成一个py文件，python执行即可。</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://biomarble.github.io/2018/03/19/Ngrok-%E9%85%8D%E7%BD%AE%E6%96%B9%E6%B3%95/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://img.discogs.com/w9vJTN3wwe3nTq7628A9o9QFb6k=/fit-in/600x597/filters:strip_icc():format(jpeg):mode_rgb():quality(90)/discogs-images/R-1752295-1430835599-7152.jpeg.jpg">
      <meta itemprop="name" content="Marble">
      <meta itemprop="description" content="Quaerendo invenietis<br>觅之，自有所获">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="BioLog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2018/03/19/Ngrok-%E9%85%8D%E7%BD%AE%E6%96%B9%E6%B3%95/" class="post-title-link" itemprop="url">Ngrok 配置方法</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2018-03-19 17:15:05" itemprop="dateCreated datePublished" datetime="2018-03-19T17:15:05+08:00">2018-03-19</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">更新于</span>
                <time title="修改时间：2021-04-16 00:00:14" itemprop="dateModified" datetime="2021-04-16T00:00:14+08:00">2021-04-16</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/IT%E6%8A%80%E6%9C%AF%E9%9A%8F%E7%AC%94/" itemprop="url" rel="index"><span itemprop="name">IT技术随笔</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <p>准备工作：域名，并做好解析（domain.org及***.**domain.org)<br>服务器端开放端口（如果有防火墙的话）</p>
<ol>
<li><p>安装go，下载编译好的包即可。设置<strong>GOROOT.GOPATH.PATH</strong>环境变量</p>
</li>
<li><p>下载ngrok源码</p>
</li>
</ol>
  <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">git clone https:&#x2F;&#x2F;github.com&#x2F;inconshreveable&#x2F;ngrok.git</span><br></pre></td></tr></table></figure>

<ol start="3">
<li><p>生成私有证书，并拷贝近assets文件夹</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cd</span> ngrok</span><br><span class="line"><span class="built_in">export</span> NGROK_DOMAIN=<span class="string">&quot;domain.org&quot;</span></span><br><span class="line">openssl genrsa -out rootCA.key 2048</span><br><span class="line">openssl req -x509 -new -nodes -key rootCA.key -subj <span class="string">&quot;/CN=<span class="variable">$NGROK_DOMAIN</span>&quot;</span> -days 5000 -out rootCA.pem</span><br><span class="line">openssl genrsa -out device.key 2048</span><br><span class="line">openssl req -new -key device.key -subj <span class="string">&quot;/CN=<span class="variable">$NGROK_DOMAIN</span>&quot;</span> -out device.csr</span><br><span class="line">openssl x509 -req -<span class="keyword">in</span> device.csr -CA rootCA.pem -CAkey rootCA.key -CAcreateserial -out device.crt -days 5000</span><br><span class="line">cp rootCA.pem assets/client/tls/ngrokroot.crt</span><br><span class="line">cp device.crt assets/server/tls/snakeoil.crt</span><br><span class="line">cp device.key assets/server/tls/snakeoil.key</span><br></pre></td></tr></table></figure></li>
<li><p>编译</p>
</li>
</ol>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">GOOS=linux GOARCH=amd64</span><br><span class="line">make release-server release-client</span><br></pre></td></tr></table></figure>

<ol start="5">
<li>打开ngrokd服务器端命令</li>
</ol>
  <figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">./bin/ngrokd -domain=<span class="string">&quot;domain.org&quot;</span>  -httpAddr=<span class="string">&quot;:80&quot;</span> -httpsAddr=<span class="string">&quot;:443&quot;</span> -tunnelAddr=<span class="string">&quot;:4443&quot;</span></span><br></pre></td></tr></table></figure>

<ol start="6">
<li><p>拷贝bin/ngrok至客户机</p>
</li>
<li><p>填写配置文件模板：</p>
</li>
</ol>
  <figure class="highlight"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">server_addr: &quot;domain.org:8082&quot;</span><br><span class="line">trust_host_root_certs: false</span><br><span class="line">tunnels:</span><br><span class="line">  http:</span><br><span class="line"> subdomain: &quot;www&quot;</span><br><span class="line"> proto:</span><br><span class="line">   http: &quot;80&quot;</span><br><span class="line">  https:</span><br><span class="line"> subdomain: &quot;www&quot;</span><br><span class="line"> proto:</span><br><span class="line">   https: &quot;80&quot;</span><br><span class="line">  ssh:</span><br><span class="line"> remote_port: 100     #映射为服务器端口，需开放</span><br><span class="line"> proto:</span><br><span class="line">   tcp: &quot;22&quot;          #待映射的客户机端口</span><br></pre></td></tr></table></figure>

<ol start="8">
<li>开机启动，设置service文件，/etc/systemd/system/ngrok.service<br> 开机自动启动   systemctl enable ngrok<br> 启动           systemctl start ngrok</li>
</ol>
  <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">[Unit]</span><br><span class="line">Description&#x3D;Ngrok Service</span><br><span class="line">Wants&#x3D;network-online.target</span><br><span class="line">After&#x3D;network.target</span><br><span class="line"></span><br><span class="line">[Service]</span><br><span class="line">Type&#x3D;simple</span><br><span class="line">ExecStart&#x3D;&#x2F;usr&#x2F;sbin&#x2F;ngrok-work -config&#x3D;&#x2F;home&#x2F;shitw&#x2F;ngrok&#x2F;LAAS.cfg -log-level&#x3D;WARNING  -log&#x3D;&#x2F;var&#x2F;log&#x2F;ngrok.log start ssh</span><br><span class="line">StandardOutput&#x3D;syslog</span><br><span class="line">StandardError&#x3D;null</span><br><span class="line">Restart&#x3D;always</span><br><span class="line">RestartSec&#x3D;30</span><br><span class="line"></span><br><span class="line">[Install]</span><br><span class="line">WantedBy&#x3D;multi-user.target</span><br></pre></td></tr></table></figure>

<ol start="9">
<li>更换ngrok端口与nginx80端口共用</li>
</ol>
<p>既然「nginx」占用了80端口，那么就在「nginx」上做文章。只要绑定ngrok.zmkj6.top 并将所有请求转发到8002端口即可。在服务器中新建/etc/nginx/conf.d/ngrock.conf文件，添加以下内容：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">upstream ngrok &#123;</span><br><span class="line">server 127.0.0.1:8002; </span><br><span class="line"><span class="comment">#此处端口要跟 启动服务端ngrok时指定的端口一致</span></span><br><span class="line">keepalive 64;</span><br><span class="line">&#125;</span><br><span class="line">server &#123;</span><br><span class="line">listen       80;</span><br><span class="line">server_name  *.ngrok.zmkj6.top;</span><br><span class="line"><span class="comment">#charset koi8-r;</span></span><br><span class="line">access_log  logs/ngrok.log;</span><br><span class="line">location / &#123;</span><br><span class="line">        proxy_set_header X-Real-IP <span class="variable">$remote_addr</span>;</span><br><span class="line">        proxy_set_header X-Forwarded-For </span><br><span class="line">        <span class="variable">$proxy_add_x_forwarded_for</span>;</span><br><span class="line">        proxy_set_header Host  <span class="variable">$http_host</span>:8002;</span><br><span class="line">       <span class="comment"># 此处端口要跟 启动服务端ngrok 时指定的端口一致</span></span><br><span class="line">        proxy_set_header X-Nginx-Proxy <span class="literal">true</span>;</span><br><span class="line">        proxy_set_header Connection <span class="string">&quot;&quot;</span>;</span><br><span class="line">        proxy_pass      http://ngrok;</span><br><span class="line">&#125;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>重新加载「nginx」</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">nginx -s reload</span><br></pre></td></tr></table></figure>
<p>此外，「ngrok」+ 「nginx」+「Docker」也可以完美实现，由于对Docker」不了解没有去折腾</p>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  

      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="http://biomarble.github.io/2017/10/01/LepMap3%E7%9A%84%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95%E8%AF%B4%E6%98%8E/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="https://img.discogs.com/w9vJTN3wwe3nTq7628A9o9QFb6k=/fit-in/600x597/filters:strip_icc():format(jpeg):mode_rgb():quality(90)/discogs-images/R-1752295-1430835599-7152.jpeg.jpg">
      <meta itemprop="name" content="Marble">
      <meta itemprop="description" content="Quaerendo invenietis<br>觅之，自有所获">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="BioLog">
    </span>
      <header class="post-header">
        <h2 class="post-title" itemprop="name headline">
          
            <a href="/2017/10/01/LepMap3%E7%9A%84%E4%BD%BF%E7%94%A8%E6%96%B9%E6%B3%95%E8%AF%B4%E6%98%8E/" class="post-title-link" itemprop="url">LepMap3的使用方法说明</a>
        </h2>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-calendar"></i>
              </span>
              <span class="post-meta-item-text">发表于</span>

              <time title="创建时间：2017-10-01 15:55:12" itemprop="dateCreated datePublished" datetime="2017-10-01T15:55:12+08:00">2017-10-01</time>
            </span>
              <span class="post-meta-item">
                <span class="post-meta-item-icon">
                  <i class="far fa-calendar-check"></i>
                </span>
                <span class="post-meta-item-text">更新于</span>
                <time title="修改时间：2021-04-16 00:00:04" itemprop="dateModified" datetime="2021-04-16T00:00:04+08:00">2021-04-16</time>
              </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="far fa-folder"></i>
              </span>
              <span class="post-meta-item-text">分类于</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/%E9%81%97%E4%BC%A0%E5%9B%BE%E8%B0%B1/" itemprop="url" rel="index"><span itemprop="name">遗传图谱</span></a>
                </span>
            </span>

          

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
          <h1 id="Lepmap3-遗传图谱构建流程"><a href="#Lepmap3-遗传图谱构建流程" class="headerlink" title="Lepmap3 遗传图谱构建流程"></a>Lepmap3 遗传图谱构建流程</h1><h2 id="Lepmap3基本流程"><a href="#Lepmap3基本流程" class="headerlink" title="Lepmap3基本流程"></a>Lepmap3基本流程</h2><p><img src="https://sourceforge.net/p/lep-map3/wiki/LM3%20Home/attachment/pipeline.png" alt="Lepmap3流程"></p>
<h2 id="运行前数据准备"><a href="#运行前数据准备" class="headerlink" title="运行前数据准备"></a>运行前数据准备</h2><h3 id="vcf文件"><a href="#vcf文件" class="headerlink" title="vcf文件"></a>vcf文件</h3><p>vcf文件必须有likelihood信息，lepmap会用likelihood信息进行分析</p>
<h3 id="pedigree文件"><a href="#pedigree文件" class="headerlink" title="pedigree文件"></a>pedigree文件</h3><p>pedigree文件包含了样本名及家系等相关信息，pedigree文件为tab分隔的文本文件，示例如下：</p>
<table>
<thead>
<tr>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
<th></th>
</tr>
</thead>
<tbody><tr>
<td>CHR</td>
<td>POS</td>
<td>F</td>
<td>F</td>
<td>F</td>
<td>F</td>
<td>F</td>
<td>F</td>
</tr>
<tr>
<td>CHR</td>
<td>POS</td>
<td>female</td>
<td>male</td>
<td>progeny_1</td>
<td>progeny_2</td>
<td>progeny_3</td>
<td>progeny_4</td>
</tr>
<tr>
<td>CHR</td>
<td>POS</td>
<td>0</td>
<td>0</td>
<td>male</td>
<td>male</td>
<td>male</td>
<td>male</td>
</tr>
<tr>
<td>CHR</td>
<td>POS</td>
<td>0</td>
<td>0</td>
<td>female</td>
<td>female</td>
<td>female</td>
<td>female</td>
</tr>
<tr>
<td>CHR</td>
<td>POS</td>
<td>2</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr>
<td>CHR</td>
<td>POS</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
</tbody></table>
<p>每行涵义如下</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">line 1：家系（Family）名称</span><br><span class="line">line 2：样本名称</span><br><span class="line">line 3：为father的名称，对应第二行第四列</span><br><span class="line">line 4：为mother的名称，对应第二行第三列</span><br><span class="line">line 5：性别，1 male, 2 female, 0 unknown</span><br><span class="line">line 6：表型，一般用不到。can be 0 for all individuals, this is not currently used</span><br></pre></td></tr></table></figure>

<h3 id="其他软件输出格式转换为Lepmap3格式"><a href="#其他软件输出格式转换为Lepmap3格式" class="headerlink" title="其他软件输出格式转换为Lepmap3格式"></a>其他软件输出格式转换为Lepmap3格式</h3><p>JoinMap loc files:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -f locsingle.awk file.loc|awk -f loc2genotypes.awk|awk -f genotypes2post.awk |java -cp Lep-MAP3/bin ... data=- ...</span><br></pre></td></tr></table></figure>

<p>Genotype assay data (in tab separated affx_genotypes.txt):</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">SNP   IND1 IND2 IND3 IND4   ...</span><br><span class="line">AX-01 AA   AB   BB   NoCall ...</span><br><span class="line">AX-02 AB   AA   AA   BB     ...</span><br><span class="line">...</span><br></pre></td></tr></table></figure>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">awk -f affx2post.awk affx_genotypes.txt|java -cp Lep-MAP3/bin ParentCall2 data=pedigree.txt posteriorFile=- ...</span><br></pre></td></tr></table></figure>




<h2 id="运行方法"><a href="#运行方法" class="headerlink" title="运行方法"></a>运行方法</h2><h4 id="ParentCall2"><a href="#ParentCall2" class="headerlink" title="ParentCall2"></a>ParentCall2</h4><p>该步骤是为了call亲本的基因型（如果已经有，可以矫正亲本基因型，如果没有，可以补缺失亲本基因型），也就是猜测亲本的基因型，<del>并且可以查看性染色体的marker</del>。 支持Grandparents信息，以及半同胞家系。</p>
<blockquote>
<p>The module ParentCall2 is used to call (possible missing or erroneous) parental genotypes and markers on sex chromosomes (XY or ZW system, assumes known offspring sex). Grandparents and half-sib familes are also supported (especially the grandparents should be provided if there is data on them).</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java ParentCall2 data=pedigree.txt vcfFile=input.vcf removeNonInformative=1|gzip &gt;data.call.gz</span><br></pre></td></tr></table></figure>

<p>Calling markers on autosomes and on X chromosome (X-)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java ParentCall2 data=... XLimit=2 ...</span><br></pre></td></tr></table></figure>

<p>Calling markers on autosomes and on Z chromosome (Z-)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java ParentCall2 data=... ZLimit=2 ...</span><br></pre></td></tr></table></figure>

<h4 id="Filtering2"><a href="#Filtering2" class="headerlink" title="Filtering2"></a>Filtering2</h4><p>该步骤可以根据偏分离（dataTolerance）和缺失率过滤标记（missingLimit）</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java Filtering2 data=data1.call dataTolerance=0.001 &gt;data_f.call</span><br></pre></td></tr></table></figure>

<h4 id="SeparateChromosomes2"><a href="#SeparateChromosomes2" class="headerlink" title="SeparateChromosomes2"></a>SeparateChromosomes2</h4><p>该步骤可以将标记进行分群</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java SeparateChromosomes2 data=data_f.call lodLimit=5 &gt;map5.txt</span><br><span class="line">java SeparateChromosomes2 data=data_f.call lodLimit=5 theta=0.05 &gt;map5.5.txt</span><br><span class="line">java SeparateChromosomes2 data=data.call lodLimit=5 distortionLod=1 &gt;map5.nofilt.txt</span><br></pre></td></tr></table></figure>

<p>(parameter numThreads utilises multiple cores)</p>
<p>The size distribution of linkage groups can be obtained like this:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sort map5.txt|uniq -c|sort -n</span><br></pre></td></tr></table></figure>



<h4 id="JoinSingles2All"><a href="#JoinSingles2All" class="headerlink" title="JoinSingles2All"></a>JoinSingles2All</h4><p>该步骤可以将上一步中未分到连锁群的标记重新插入到已有的连锁群中</p>
<blockquote>
<p>The JoinSingles2All module assigns singular markers to existing LGs by computing LOD scores between each single marker and markers from the existing LGs. It needs a map file as input parameter and then it outputs a new map file with possible more markers assigned to LGs.</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ JoinSingles2All map=map5.txt data=data_f.call lodLimit=4 &gt;map5_js.txt</span><br><span class="line">zcat data_f.call.gz|java -cp bin/ JoinSingles2All map=map5.txt data=- lodLimit=3 lodDifference=2 &gt;map5_js.txt</span><br></pre></td></tr></table></figure>

<p>(parameter numThreads utilises muliple cores)</p>
<p>(iterated joinSingles2All yields same result as iterating it until no markers can be added)</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ JoinSingles2All map=map5.txt data=data_f.call lodLimit=4 iterate=1 &gt;map5_js_iterated.txt</span><br></pre></td></tr></table></figure>

<p>The size distribution of linkage groups can be obtained like this:</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cut -f 1 map5_js.txt|sort|uniq -c|sort -n</span><br></pre></td></tr></table></figure>

<h4 id="OrderMarkers2"><a href="#OrderMarkers2" class="headerlink" title="OrderMarkers2"></a>OrderMarkers2</h4><p>对每个连锁群内的标记进行排序，以最大化likelihood</p>
<blockquote>
<p>OrderMarkers2 orders the markers within each LG by maximizing the likelihood of the data given the order. Either a map file (map=map_file) or output of previous run of OrderMarkers2 (evaluateOrder=order_file) have to be provided (only the first column (the marker name or lg name) is read from these files).</p>
</blockquote>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ OrderMarkers2 map=map5_js.txt data=data_f.call chromosome=1 &gt;order1.txt</span><br><span class="line">java -cp bin/ OrderMarkers2 evaluateOrder=order1.txt data=data_f.call &gt;order1.2.txt</span><br></pre></td></tr></table></figure>

<p>已知标记顺序，计算遗传距离</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ OrderMarkers2 evaluateOrder=order2.txt data=data_f.call improveOrder=0 &gt;order1.3.txt</span><br></pre></td></tr></table></figure>

<p>已知标记顺序，计算中性图距离</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ OrderMarkers2 evaluateOrder=order1.txt data=data_f.call improveOrder=0 sexAveraged=1 &gt;order1.SA.txt</span><br></pre></td></tr></table></figure>

<p>无交叉减数分裂achiasmatic meiosis (父本中无重组交换):</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ OrderMarkers2 map=map.txt data=data_f.call recombination1=0</span><br></pre></td></tr></table></figure>

<p>无交叉减数分裂achiasmatic meiosis (母本中无重组交换):</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ OrderMarkers2 map=map.txt data=data_f.call recombination2=0</span><br></pre></td></tr></table></figure>

<p>分染色体排序</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin/ OrderMarkers2 map=mapBig.txt data=dataBig.call chromosome=1 &gt;order1.1.txt</span><br><span class="line">...</span><br><span class="line">java -cp bin/ OrderMarkers2 map=mapBig.txt data=dataBig.call chromosome=N &gt;orderN.1.txt</span><br></pre></td></tr></table></figure>

<h4 id="LMPlot"><a href="#LMPlot" class="headerlink" title="LMPlot"></a>LMPlot</h4><p>绘制点图<br>for every chromosome :</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin&#x2F; OrderMarkers2 map&#x3D;map.txt data&#x3D;data_f.call outputPhasedData&#x3D;1 chromosome&#x3D;1 &gt;order_wp_1.txt</span><br><span class="line">java -cp bin&#x2F; LMPlot order_wp_1.txt &gt;order_wp_1.dot</span><br><span class="line">xdot.py order_wp_1.dot</span><br></pre></td></tr></table></figure>

<p>or:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">java -cp bin&#x2F; OrderMarkers2 evaluateOrder&#x3D;order1.txt data&#x3D;data_f.call outputPhasedData&#x3D;1 &gt;order_wp_1.txt</span><br><span class="line">java -cp bin&#x2F; LMPlot order_wp_1.txt &gt;order_wp_1.dot</span><br><span class="line">xdot.py order_wp_1.dot</span><br></pre></td></tr></table></figure>

      
    </div>

    
    
    
      <footer class="post-footer">
        <div class="post-eof"></div>
      </footer>
  </article>
  
  
  


  



          </div>
          

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt="Marble"
      src="https://img.discogs.com/w9vJTN3wwe3nTq7628A9o9QFb6k=/fit-in/600x597/filters:strip_icc():format(jpeg):mode_rgb():quality(90)/discogs-images/R-1752295-1430835599-7152.jpeg.jpg">
  <p class="site-author-name" itemprop="name">Marble</p>
  <div class="site-description" itemprop="description">Quaerendo invenietis<br>觅之，自有所获</div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives/">
        
          <span class="site-state-item-count">3</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
          
        <span class="site-state-item-count">3</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
          
        <span class="site-state-item-count">6</span>
        <span class="site-state-item-name">标签</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/biomarble" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;biomarble" rel="noopener" target="_blank"><i class="fab fa-github fa-fw"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="mailto:shitongwei@gmail.com" title="E-Mail → mailto:shitongwei@gmail.com" rel="noopener" target="_blank"><i class="fa fa-envelope fa-fw"></i>E-Mail</a>
      </span>
      <span class="links-of-author-item">
        <a href="https://music.163.com/#/user/home?id=247828756" title="网易云音乐 → https:&#x2F;&#x2F;music.163.com&#x2F;#&#x2F;user&#x2F;home?id&#x3D;247828756" rel="noopener" target="_blank"><i class="fa fa-play-circle fa-fw"></i>网易云音乐</a>
      </span>
  </div>



      </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 2018 – 
  <span itemprop="copyrightYear">2021</span>
  <span class="with-love">
    <i class="fa fa-dna"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Marble</span>
</div>

        








      </div>
    </footer>
  </div>

  
  <script src="/lib/anime.min.js"></script>
  <script src="/lib/velocity/velocity.min.js"></script>
  <script src="/lib/velocity/velocity.ui.min.js"></script>

<script src="/js/utils.js"></script>

<script src="/js/motion.js"></script>


<script src="/js/schemes/pisces.js"></script>


<script src="/js/next-boot.js"></script>




  















  

  

</body>
</html>
